### ISSUE TITLE
Feature Request: Support `componentDidCatch` in `renderToStaticMarkup`

### ISSUE THREAD

**Issue by AndrewLeedham** [facebook/react#21500]
It is mentioned in https://github.com/facebook/react/issues/10442 that `renderToString` and `renderToNodeStream` are not supported due to streaming limitations which makes sense. However, `renderToStaticMarkup` is listed in the docs as being for static page generation.

> This is useful if you want to use React as a simple static page generator, as stripping away the extra attributes can save some bytes.

Therefore, would it be possible to support `componentDidCatch`/Error Boundaries when using `renderToStaticMarkup`.

My use-case is using React to build static HTML pages server-side and wanting to scope errors to components so failed components are removed from the page, but the rest of the page is still rendered.

I created a CodeSandbox to illustrate the current behaviour:
https://codesandbox.io/s/rendertostaticmarkup-errorboundary-ec1ep

**Comment by LXSMNSYC** [comment]
I believe `getDerivedStateFromError` is a better workaround, specially since it works in SSR. However, there's a lack of component stack info.

**Comment by bvaughn** [comment]
`componentDidCatch` is a method called during React's _commit phase_. There's no commit phase on the server. (The commit phase is when it's safe to e.g. read layout or adjust DOM elements.) Since there's no actual DOM to read from with `renderToStaticMarkup`, calling commit phase methods would likely lead to lots of errors.

`getDerivedStateFromError` is called during React's _render phase_ so it would be more appropriate in this case.

**Comment by AndrewLeedham** [comment]
My example uses `getDerivedStateFromError` but the `renderToStaticMarkup` call still fails? Is the suggestion that the feature request should be for `getDerivedStateFromError` to work in `renderToStaticMarkup`, or are you saying it should currently be working?

**Comment by bvaughn** [comment]
No. I don't believe `renderToStaticMarkup` supports error boundaries (including `getDerivedStateFromError`) but that method would be the more appropriate API to use since there's no commit phase when rendering static markup.
